internal/util/crypto.go

```diff
@@ -57,56 +52,111 @@ const (
-	DeletePassphrase(key string) error
-	GetID() string
-}
+var (
+	// ErrDEKStoreNotFound is an error that is returned when the DEKStore
+	// has not been configured for the volumeID in the KMS instance.
+	ErrDEKStoreNotFound = errors.New("DEKStore not found")
+
+	// ErrDEKStoreNeeded is an indication that gets returned with
+	// NewVolumeEncryption when the KMS does not include support for the
+	// DEKStore interface.
+	ErrDEKStoreNeeded = errors.New("DEKStore required, use " +
+		"VolumeEncryption.SetDEKStore()")
+)
+
+type VolumeEncryption struct {
+	KMS EncryptionKMS
 
-// SecretsKMS is default KMS implementation that means no KMS is in use.
-type SecretsKMS struct {
-	passphrase string
+	// dekStore that will be used, this can be the EncryptionKMS or a
+	// different object implementing the DEKStore interface.
+	dekStore DEKStore
 }
 
-func initSecretsKMS(secrets map[string]string) (EncryptionKMS, error) {
-	passphraseValue, ok := secrets[encryptionPassphraseKey]
-	if !ok {
-		return nil, errors.New("missing encryption passphrase in secrets")
+// NewVolumeEncryption creates a new instance of VolumeEncryption and
+// configures the DEKStore. If the KMS does not provide a DEKStore interface,
+// the VolumeEncryption will be created *and* a ErrDEKStoreNeeded is returned.
+// Callers that receive a ErrDEKStoreNeeded error, should use
+// VolumeEncryption.SetDEKStore() to configure an alternative storage for the
+// DEKs.
+func NewVolumeEncryption(kms EncryptionKMS) (*VolumeEncryption, error) {
+	ve := &VolumeEncryption{KMS: kms}
+
+	if kms.requiresDEKStore() == DEKStoreIntegrated {
+		dekStore, ok := kms.(DEKStore)
+		if !ok {
+			return nil, fmt.Errorf("KMS %T does not implement the "+
+				"DEKStore interface", kms)
+		}
+
+		ve.dekStore = dekStore
+		return ve, nil
 	}
-	return SecretsKMS{passphrase: passphraseValue}, nil
+
+	return ve, ErrDEKStoreNeeded
 }
 
-// Destroy frees all used resources.
-func (kms SecretsKMS) Destroy() {
-	// nothing to do
+func (ve *VolumeEncryption) SetDEKStore(dekStore DEKStore) {
+	ve.dekStore = dekStore
 }
 
-// GetPassphrase returns passphrase from Kubernetes secrets.
-func (kms SecretsKMS) GetPassphrase(key string) (string, error) {
-	return kms.passphrase, nil
+func (ve *VolumeEncryption) Destroy() {
+	ve.KMS.Destroy()
 }
 
-// SavePassphrase does nothing, as there is no passphrase per key (volume), so
-// no need to store is anywhere.
-func (kms SecretsKMS) SavePassphrase(key, value string) error {
-	return nil
+func (ve *VolumeEncryption) RemoveDEK(volumeID string) error {
+	if ve.dekStore == nil {
+		return ErrDEKStoreNotFound
+	}
+
+	return ve.dekStore.RemoveDEK(volumeID)
 }
 
-// DeletePassphrase is doing nothing as no new passphrases are saved with
-// SecretsKMS.
-func (kms SecretsKMS) DeletePassphrase(key string) error {
-	return nil
+// EncryptionKMS provides external Key Management System for encryption
+// passphrases storage.
+type EncryptionKMS interface {
```

Comment:
A wide guess here, do we need a registration method to be clubbed with EncryptionKMS interface ? so that the registration or connection to KMS can be implemented for a KMS with that.

Reply:
I thought about that too. It would be nice to have, but not something required for this PR. Maybe create an issue for such a cleanup?

Reply:
sure. In one way keeping the KMS methods with the connection/registration handle would be good. will open an issue for the same.